File: lib/features/routes/domain/hike_duration_estimate.dart
import 'dart:math' as math;

class HikeSegmentInput {
  const HikeSegmentInput({
    required this.distanceM,
    this.deltaAltM,
  });

  final double distanceM;

  final int? deltaAltM;
}

class HikeDurationResult {
  const HikeDurationResult({
    required this.distanceKm,
    required this.ascentM,
    required this.descentM,
    required this.durationH,
    this.durationHint,
  });

  final double distanceKm;
  final int ascentM;
  final int descentM;
  final double durationH;
  final String? durationHint;
}

class HikeDurationEstimate {
  HikeDurationEstimate._();

  static const double flatSpeedKmh = 4.0;
  static const double ascentMinPer100m = 10.0;
  static const double descentMinPer100mGentle = 4.0;
  static const double descentMinPer100mSteep = 9.0;
  static const double steepGradePercent = 18.0;
  static const double fatigueEffortThreshold = 8.0;
  static const double maxFatigueMultiplier = 1.35;

  static HikeDurationResult fromSegments(List<HikeSegmentInput> segments) {
    var totalMeters = 0.0;
    var ascent = 0;
    var descent = 0;
    var totalHours = 0.0;
    var cumulativeKm = 0.0;
    var cumulativeAscent = 0.0;
    var hadAltitude = false;
    var hadSteepDescent = false;
    var fatigueApplied = false;

    for (final seg in segments) {
      final distKm = seg.distanceM / 1000.0;
      if (distKm <= 0 && (seg.deltaAltM ?? 0) == 0) continue;

      totalMeters += seg.distanceM;
      final delta = seg.deltaAltM;
      if (delta != null) {
        hadAltitude = true;
        if (delta > 0) ascent += delta;
        if (delta < 0) descent += -delta;
      }

      final fatigue = _fatigueMultiplier(cumulativeKm, cumulativeAscent);
      if (fatigue > 1.01) fatigueApplied = true;

      final segHours = _segmentHours(distKm, delta);
      if (delta != null &&
          delta < 0 &&
          distKm > 0.001 &&
          (-delta / (distKm * 1000)) * 100 >= steepGradePercent) {
        hadSteepDescent = true;
      }

      totalHours += segHours * fatigue;
      cumulativeKm += distKm;
      if (delta != null && delta > 0) cumulativeAscent += delta;
    }

    final km = totalMeters / 1000.0;
    if (km == 0) {
      return const HikeDurationResult(
        distanceKm: 0,
        ascentM: 0,
        descentM: 0,
        durationH: 0,
      );
    }

    if (!hadAltitude) {
      totalHours = _distanceOnlyWithFatigue(km);
    }

    final hints = <String>[];
    if (hadAltitude && descent > 0) {
      hints.add('враховано спуск $descent м');
    }
    if (fatigueApplied) {
      hints.add('додано час на втомливість на довгих ділянках');
    }
    if (hadSteepDescent) {
      hints.add('крутий спуск уповільнює темп');
    }

    return HikeDurationResult(
      distanceKm: double.parse(km.toStringAsFixed(2)),
      ascentM: ascent,
      descentM: descent,
      durationH: double.parse(totalHours.toStringAsFixed(1)),
      durationHint: hints.isEmpty ? null : '${hints.join('; ')}.',
    );
  }

  static double _segmentHours(double distKm, int? deltaAltM) {
    if (distKm <= 0) {
      if (deltaAltM == null || deltaAltM == 0) return 0;
      final climb = deltaAltM > 0 ? deltaAltM : -deltaAltM;
      final minPer100 = deltaAltM > 0 ? ascentMinPer100m : descentMinPer100mSteep;
      return (climb / 100.0) * (minPer100 / 60.0);
    }

    final flatH = distKm / flatSpeedKmh;
    final delta = deltaAltM ?? 0;

    if (delta > 0) {
      final ascentH = (delta / 100.0) * (ascentMinPer100m / 60.0);
      final grade = (delta / (distKm * 1000)) * 100;
      final steepFactor = grade > 25
          ? 1.18
          : grade > 15
              ? 1.10
              : 1.0;
      return (flatH + ascentH) * steepFactor;
    }

    if (delta < 0) {
      final loss = -delta;
      final grade = (loss / (distKm * 1000)) * 100;
      final minPer100 = grade >= steepGradePercent
          ? descentMinPer100mSteep
          : descentMinPer100mGentle;
      final descentH = (loss / 100.0) * (minPer100 / 60.0);
      final downhillFlatH = distKm / 3.3;
      return math.max(flatH, downhillFlatH) + descentH;
    }

    return flatH;
  }

  static double _fatigueMultiplier(double cumulativeKm, double cumulativeAscentM) {
    final effort = cumulativeKm + cumulativeAscentM / 400.0;
    if (effort <= fatigueEffortThreshold) return 1.0;
    final excess = effort - fatigueEffortThreshold;
    return (1.0 + 0.028 * excess).clamp(1.0, maxFatigueMultiplier);
  }

  static double _distanceOnlyWithFatigue(double km) {
    var hours = 0.0;
    var walked = 0.0;
    const stepKm = 0.5;
    while (walked < km) {
      final chunk = math.min(stepKm, km - walked);
      hours += (chunk / flatSpeedKmh) *
          _fatigueMultiplier(walked, 0);
      walked += chunk;
    }
    return hours;
  }
}

File: supabase/functions/_shared/hike_duration_estimate.ts


export type HikeSegmentInput = {
  distance_m: number;
  delta_alt_m?: number | null;
};

export type HikeDurationResult = {
  distance_km: number;
  ascent_m: number;
  descent_m: number;
  duration_h: number;
};

const FLAT_SPEED_KMH = 4.0;
const ASCENT_MIN_PER_100M = 10.0;
const DESCENT_MIN_GENTLE = 4.0;
const DESCENT_MIN_STEEP = 9.0;
const STEEP_GRADE_PERCENT = 18;
const FATIGUE_EFFORT_THRESHOLD = 8;
const MAX_FATIGUE = 1.35;

function fatigueMultiplier(cumulativeKm: number, cumulativeAscentM: number): number {
  const effort = cumulativeKm + cumulativeAscentM / 400;
  if (effort <= FATIGUE_EFFORT_THRESHOLD) return 1;
  const excess = effort - FATIGUE_EFFORT_THRESHOLD;
  return Math.min(MAX_FATIGUE, 1 + 0.028 * excess);
}

function segmentHours(distKm: number, deltaAltM: number | null | undefined): number {
  if (distKm <= 0) {
    if (deltaAltM == null || deltaAltM === 0) return 0;
    const climb = Math.abs(deltaAltM);
    const minPer100 = deltaAltM > 0 ? ASCENT_MIN_PER_100M : DESCENT_MIN_STEEP;
    return (climb / 100) * (minPer100 / 60);
  }

  const flatH = distKm / FLAT_SPEED_KMH;
  const delta = deltaAltM ?? 0;

  if (delta > 0) {
    const ascentH = (delta / 100) * (ASCENT_MIN_PER_100M / 60);
    const grade = (delta / (distKm * 1000)) * 100;
    const steepFactor = grade > 25 ? 1.18 : grade > 15 ? 1.1 : 1;
    return (flatH + ascentH) * steepFactor;
  }

  if (delta < 0) {
    const loss = -delta;
    const grade = (loss / (distKm * 1000)) * 100;
    const minPer100 = grade >= STEEP_GRADE_PERCENT
      ? DESCENT_MIN_STEEP
      : DESCENT_MIN_GENTLE;
    const descentH = (loss / 100) * (minPer100 / 60);
    const downhillFlatH = distKm / 3.3;
    return Math.max(flatH, downhillFlatH) + descentH;
  }

  return flatH;
}

function distanceOnlyWithFatigue(km: number): number {
  let hours = 0;
  let walked = 0;
  const stepKm = 0.5;
  while (walked < km) {
    const chunk = Math.min(stepKm, km - walked);
    hours += (chunk / FLAT_SPEED_KMH) * fatigueMultiplier(walked, 0);
    walked += chunk;
  }
  return hours;
}

export function estimateHikeDurationFromSegments(
  segments: HikeSegmentInput[],
): HikeDurationResult {
  let totalMeters = 0;
  let ascent = 0;
  let descent = 0;
  let totalHours = 0;
  let cumulativeKm = 0;
  let cumulativeAscent = 0;
  let hadAltitude = false;

  for (const seg of segments) {
    const distKm = seg.distance_m / 1000;
    if (distKm <= 0 && (seg.delta_alt_m ?? 0) === 0) continue;

    totalMeters += seg.distance_m;
    const delta = seg.delta_alt_m;
    if (delta != null) {
      hadAltitude = true;
      if (delta > 0) ascent += delta;
      if (delta < 0) descent += -delta;
    }

    const fatigue = fatigueMultiplier(cumulativeKm, cumulativeAscent);
    totalHours += segmentHours(distKm, delta) * fatigue;
    cumulativeKm += distKm;
    if (delta != null && delta > 0) cumulativeAscent += delta;
  }

  const km = totalMeters / 1000;
  if (km === 0) {
    return { distance_km: 0, ascent_m: 0, descent_m: 0, duration_h: 0 };
  }

  if (!hadAltitude) {
    totalHours = distanceOnlyWithFatigue(km);
  }

  return {
    distance_km: Math.round(km * 100) / 100,
    ascent_m: ascent,
    descent_m: descent,
    duration_h: Math.round(totalHours * 10) / 10,
  };
}

export function estimateHikeDurationFromPoints(
  points: { lat: number; lon: number; altitude_m?: number | null }[],
  haversineM: (a: { lat: number; lon: number }, b: { lat: number; lon: number }) => number,
): HikeDurationResult {
  const coords = points.filter((p) => Number.isFinite(p.lat) && Number.isFinite(p.lon));
  const segments: HikeSegmentInput[] = [];
  for (let i = 1; i < coords.length; i++) {
    const prev = coords[i - 1];
    const cur = coords[i];
    const distance_m = haversineM(
      { lat: prev.lat, lon: prev.lon },
      { lat: cur.lat, lon: cur.lon },
    );
    let delta_alt_m: number | null = null;
    const pa = prev.altitude_m;
    const ca = cur.altitude_m;
    if (pa != null && ca != null) delta_alt_m = ca - pa;
    segments.push({ distance_m, delta_alt_m });
  }
  return estimateHikeDurationFromSegments(segments);
}
